ext/http/network_buffered_stream.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use bytes::Bytes;
+use deno_core::futures::future::poll_fn;
+use deno_core::futures::ready;
+use std::io;
+use std::mem::MaybeUninit;
+use std::pin::Pin;
+use std::task::Poll;
+use tokio::io::AsyncRead;
+use tokio::io::AsyncWrite;
+use tokio::io::ReadBuf;
+
+const MAX_PREFIX_SIZE: usize = 256;
+
+pub struct NetworkStreamPrefixCheck<S: AsyncRead + Unpin> {
+  io: S,
+  prefix: &'static [u8],
+  buffer: [MaybeUninit<u8>; MAX_PREFIX_SIZE * 2],
+}
+
+impl<S: AsyncRead + Unpin> NetworkStreamPrefixCheck<S> {
+  pub fn new(io: S, prefix: &'static [u8]) -> Self {
+    debug_assert!(prefix.len() < MAX_PREFIX_SIZE);
+    Self {
+      io,
+      prefix,
+      buffer: [MaybeUninit::<u8>::uninit(); MAX_PREFIX_SIZE * 2],
+    }
+  }
+
+  // Returns a [`NetworkBufferedStream`], rewound with the bytes we read to determine what
+  // type of stream this is.
+  pub async fn match_prefix(
+    self,
+  ) -> io::Result<(bool, NetworkBufferedStream<S>)> {
+    let mut buffer = self.buffer;
+    let mut readbuf = ReadBuf::uninit(&mut buffer);
+    let mut io = self.io;
+    let prefix = self.prefix;
+    loop {
+      enum State {
+        Unknown,
+        Matched,
+        NotMatched,
+      }
+
+      let state = poll_fn(|cx| {
+        let filled_len = readbuf.filled().len();
+        let res = ready!(Pin::new(&mut io).poll_read(cx, &mut readbuf));
+        if let Err(e) = res {
+          return Poll::Ready(Err(e));
+        }
+        let filled = readbuf.filled();
+        let new_len = filled.len();
+        if new_len == filled_len {
+          // Empty read, no match
+          return Poll::Ready(Ok(State::NotMatched));
+        } else if new_len < prefix.len() {
+          // Read less than prefix, make sure we're still matching the prefix (early exit)
+          if !prefix.starts_with(filled) {
+            return Poll::Ready(Ok(State::NotMatched));
+          }
+        } else if new_len >= prefix.len() {
+          // We have enough to determine
+          if filled.starts_with(prefix) {
+            return Poll::Ready(Ok(State::Matched));
+          } else {
+            return Poll::Ready(Ok(State::NotMatched));
+          }
+        }
+
+        Poll::Ready(Ok(State::Unknown))
+      })
+      .await?;
+
+      match state {
+        State::Unknown => continue,
+        State::Matched => {
+          let initialized_len = readbuf.filled().len();
+          return Ok((
+            true,
+            NetworkBufferedStream::new(io, buffer, initialized_len),
+          ));
+        }
+        State::NotMatched => {
+          let initialized_len = readbuf.filled().len();
+          return Ok((
+            false,
+            NetworkBufferedStream::new(io, buffer, initialized_len),
+          ));
+        }
+      }
+    }
+  }
+}
+
+pub struct NetworkBufferedStream<S: AsyncRead + Unpin> {
```

Comment:
Please add a brief doc comment.

Tangent: I copied the code into ChatGPT and asked it what it does, and it's pretty good. Maybe we don't need doc comments anymore soon :laughing:

> The NetworkBufferedStream is a custom wrapper around an asynchronous stream that implements AsyncRead and AsyncWrite. It is designed to provide additional buffering functionality to the wrapped stream. The primary use case for this struct is when you want to read a small amount of data from the beginning of a stream, process it, and then continue reading the rest of the stream.
> 
> The struct is generic over the type parameter S, which represents the underlying asynchronous stream that must implement the AsyncRead and Unpin traits.
>
> It has the following fields:
>
> - io: The underlying asynchronous stream.
> - initialized_len: The length of the valid data in the prefix buffer.
> - prefix_offset: The current position in the prefix buffer.
> - prefix: A fixed-size buffer used to store the initial data read from the stream.
> - prefix_read: A boolean flag to indicate if the prefix buffer has been fully read.
> 
> The struct provides the following methods:
> 
> - new: Constructs a new NetworkBufferedStream instance, taking the underlying stream, a prefix buffer, and the initialized_len as arguments.
> - current_slice: Returns a byte slice of the initialized data in the prefix buffer, from the current prefix_offset to the initialized_len.
> - into_inner: Consumes the NetworkBufferedStream and returns the underlying stream and the remaining data in the prefix buffer as a Bytes object.
> - poll_read: An implementation of the AsyncRead trait, which reads data from the prefix buffer first and then continues reading from the underlying stream once the prefix buffer is empty.
> - poll_write, poll_flush, poll_shutdown, is_write_vectored, poll_write_vectored: Implementations of the AsyncWrite trait, which delegate their functionality to the underlying stream.
>
> In summary, NetworkBufferedStream is a utility wrapper around an asynchronous stream that provides additional buffering functionality. It can be used when you need to read and process a small amount of data from the beginning of a stream before continuing to read the rest of the stream.

Reply:
Hah, not bad. It missed the fact that it's designed to look for a given prefix with the minimal amount of reading, and abort as quickly as possible without buffering excessive amounts of data. I'll create a second PR with this.